Add combined RCA blog post for July 2026 incidents (#1790)
inngest/website: content/blog/2026-07-23-incident-rca-july-2026.md

These fallback values changed how our internal services routed traffic and scheduled work, which in turn degraded our internal network and disrupted function execution and run scheduling. Separately, a code regression in the same release caused a subset of retried requests to fail. We rolled back the release, which coincided with the third-party provider restoring service, and the majority of impact was resolved.

Function execution was degraded for 60 minutes (16:45–17:45 UTC) and run scheduling for 125 minutes (16:10–18:15 UTC).

### What happened

Connectivity to the third-party service that supplies our feature-flag values degraded. Our services were unable to fetch flag values, both on startup and via streaming updates, so all affected feature flags reverted to their default states.

When we rolled out a new release that day, newly started containers were unable to retrieve feature-flag values and fell back to default configuration. Two consequences followed:

> - A subset of internal traffic switched from its optimized transport path to a legacy path, which overloaded our internal network. Between 16:45 and 17:47 UTC, network performance was severely degraded: DNS query volume increased roughly tenfold, DNS lookup latency spiked, and connection setup times rose sharply. Traffic on a key internal path effectively stalled, breaking function execution.

Reviewer: we could be more specific to say traffic shifted from gRPC to the legacy internal http traffic which caused these issues. This was left in the code as per-customer opt out if the system caused issues.

Sometimes in cases like this, being specific is nice as the audience is technical and it also indicates some of the workings of our system where we did use gRPC to help this issue and it was a regression.

- Another subsystem was configured incorrectly, leading to retries in run scheduling for a subset of customers using debounce.

In parallel, the same release contained a regression affecting certain idempotent requests on the scheduling path. During retries, this caused a subset of events to repeatedly fail until the release was rolled back.

Customer-facing impact:

> - Function execution was delayed due to impaired internal connectivity. A subset of steps may have failed between 16:45 and 17:45 UTC. These should have retried, except for function runs with disabled or exhausted retries.

Reviewer: Let's remove "may have" (they did) - may have seems like we aren't sure that it happened. We should be precise with impact.

| ----- | ---------------------------------------------------------------------------------------------------------------------- |
| 16:11 | Customer impact began. A subset of events began failing on retry due to the request regression on the scheduling path. |
| 16:19 | First feature-flag initialization errors observed as containers failed to fetch flag values.                           |
| 16:20 | Scheduling errors and retries started for a subset of customers.                                                       |
| 16:45 | Internal network performance began degrading severely as traffic shifted to the legacy path.                           |
| 17:40 | Release rolled back. Retries on the affected request path stopped failing.                                             |
| 17:41 | Internal gateway metrics began returning toward normal.                                                                |
| 17:45 | Function execution impact ended.                                                                                       |
| 17:47 | Internal network performance recovered.                                                                                |

> | 17:57 | Rollback fully propagated; status moved to Monitoring (Severity upgraded to Critical).                                 |

Reviewer: I kind of agree w/ gester - maybe the text in the parens aren't needed

- **Improve observability.** Improve run-scheduling observability to detect events that did not produce runs, implement event auditing (including when an event was intentionally debounced or rate-limited).
- **Improve resilience of internal tooling and configuration.** Reduce the blast radius of the internal network-translation layer, harden deployment tooling so rollbacks are not delayed during incidents, and complete planned upgrades of the internal gateway infrastructure.

## Incident 2 — July 16: Increased execution latency and checkpointing errors

### Summary

On July 16, 2026, Inngest experienced an incident that caused increased execution latency and elevated errors for customers.

The issue occurred while we were preparing our state store infrastructure for upcoming upgrades. As part of that work, we had enabled an upgrade-readiness state store path behind a feature flag. That path depended on an internal state store connection during startup. While the path was covered by monitoring, it was tied to a medium-priority alert and a lower SLO than was appropriate for the customer impact this failure mode could create.

> When that dependency began timing out, newly restarted state store proxy instances were unable to become ready. As more instances restarted, available proxy capacity dropped until the checkpointing API could no longer serve traffic reliably. We resolved the incident by disabling the upgrade-readiness state store path and restarting the affected proxy instances.

Reviewer: Not sure what the "proxy" is here as a reader. Maybe we should include that in the above paragraph along the lines of "a new proxy now routes traffic between our legacy and future databases."

This did not affect traffic immediately, but over time the available proxy capacity degraded as instances restarted and failed to rejoin service. Once capacity dropped far enough, checkpointing requests from API services and some executors began failing.

For functions using checkpointing, SDKs were unable to persist intermediate step progress through the checkpointing API. When checkpoint writes failed, SDKs fell back to returning completed step results through the normal async response path. This fallback is designed to preserve completed step output where possible, but it is less efficient and more sensitive to retries than the checkpointed path.

Customer-facing impact:

- Customers experienced increased execution latency and elevated execution errors.
- Checkpointing requests failed for affected functions during the incident window.

> - SDKs fell back to the normal async response path when checkpoint writes failed, which increased retries and latency.

Reviewer: ```suggestion
- SDKs fell back to the async response path to prevent data loss when checkpoint writes failed. This increased retries and latency.
```

We have also added safeguards before continuing this upgrade work:

- Removed the startup dependency that allowed one unhealthy state store path to prevent proxy instances from serving traffic.
- Moved monitoring for checkpointing API errors, proxy retry rates, and upgrade-readiness state store paths to higher-priority alerts with SLOs that better reflect customer impact.
- Added coverage for checkpointing API failures so customer-impacting errors are detected more directly.
- Added timeout-at-startup coverage to our upgrade validation so proxy restarts are tested against the same failure mode that occurred here.

## Incident 3 — July 16: Delayed execution metrics

### Summary

On July 16, 2026, starting at approximately 15:45 UTC, publishing of customer execution metrics to our metrics pipeline began to fail. During the incident window, dashboards showed no function execution or throughput metrics for affected customers. Function execution itself was not affected.

> The underlying issue was a gap between our live infrastructure configuration and its declarative definitions. The metrics publishing path had been introduced earlier this year as part of scaling our metrics pipeline, and the associated message-broker topic and access permissions were live in production ahead of their declarative definitions being completed. During maintenance to prepare the cluster for upcoming capacity work, broker resources were reconciled from those definitions, which unintentionally removed the not-yet-declared permissions. Metrics publishing began failing with authorization errors immediately afterward.

Reviewer: ```suggestion
The underlying issue was drift between our live infrastructure configuration and its declarative definitions. The metrics publishing path had been introduced earlier this year as part of scaling our metrics pipeline, and the associated message-broker topic and access permissions were live in production ahead of their declarative definitions being completed. During maintenance to prepare the cluster for upcoming capacity work, broker resources were reconciled from those definitions, which unintentionally removed the not-yet-declared permissions. Metrics publishing began failing with authorization errors immediately afterward.
```

### Timeline (UTC, July 16, 2026)

| Time  | Event                                                                             |
| ----- | --------------------------------------------------------------------------------- |
| 15:45 | Metrics publishing errors started; customer impact began.                         |
| 17:44 | Incident declared and investigation started.                                      |
| 17:55 | Topic and access permission definitions reapplied; affected services restarted.   |
| 19:05 | Publishing confirmed fully recovered and error volume at zero; incident resolved. |

### Root cause

The primary cause was infrastructure configuration drift: the message-broker topic and access permissions for the execution metrics path were live in production ahead of their declarative infrastructure definitions being completed. When broker resources were later reconciled from those definitions during maintenance, the permissions were removed and publishing failed.

Contributing factors:

- Execution metrics had recently moved to being published through the message broker rather than written directly to the analytics store, increasing the impact of this path.

> - We had not yet set up alerting or monitoring on this publishing path, which delayed detection.

Reviewer: ```suggestion
- The publishing path had insufficient monitoring and alerting, which delayed detection.
```
Reads more factual of the factor rather than what we didn't do which reads like a follow up

### Follow-ups

- Completed the declarative definitions for the metrics topic and access permissions, making them the source of truth.
- Added CI safeguards to detect drift between live broker configuration and the declarative definitions.
- Adding alerting on metrics publishing errors so failures on this path are detected immediately.
- Strengthened our rollout process so new infrastructure resources are declared and reviewed as part of their initial rollout.

## Incident 4 — July 23: Event processing outage and delayed function execution

### Summary

> On July 23, 2026, starting at 14:03 UTC, the service responsible for processing incoming events and scheduling function runs stopped processing. Event processing was fully unavailable for approximately 37 minutes, and function execution was delayed until backlogs were processed at approximately 15:12 UTC.

Reviewer: ```suggestion
On July 23, 2026, starting at 14:03 UTC, the service responsible for processing incoming events and scheduling function runs stopped processing. Event processing was fully unavailable for approximately 37 minutes, and function execution was delayed until backlogs were fully processed at approximately 15:12 UTC.
```

### What happened

As part of ongoing infrastructure work, we had introduced a new storage backend for one message-broker topic used by a new event lifecycle publishing path. On July 23, that storage backend stopped allocating new capacity on the affected volume, even though free space was available elsewhere in the cluster. As a result, the brokers backing this topic became unhealthy and clients timed out connecting to them.

The event-processing service treated a connection to this publishing path as a startup requirement. As instances restarted, they failed to connect and could not become ready, which led to a full halt of event processing and run scheduling.

Automated alerts fired within minutes and the on-call team responded immediately. Once the root cause was identified, we disabled the event lifecycle publishing path, which isolated the affected topic and allowed the event-processing service to start up and resume processing. We then scaled the affected services to work through the backlog, and held new releases until throughput returned to normal.

Customer-facing impact:

> - New events were not processed and new function runs were not scheduled between 14:03 and 14:40 UTC (14:53 UTC for a subset of customers on dedicated infrastructure).

Reviewer: ```suggestion
- New events experienced delayed processing and new function runs were not scheduled between 14:03 and 14:40 UTC (14:53 UTC for a subset of customers on dedicated infrastructure).
```

| 14:53 | Event processing resumed for the remaining customers on dedicated infrastructure.                          |
| 15:10 | Affected path fully isolated; status moved to Monitoring.                                                  |
| 15:12 | Backlog fully processed; run scheduling latency back to normal.                                            |

Duration of impact:

- **Event processing outage:** ~37 minutes (14:03–14:40 UTC).
- **Delayed run scheduling until full recovery:** ~69 minutes (14:03–15:12 UTC).

### Root cause

The primary cause was a capacity-allocation failure in a newly introduced storage backend for one message-broker topic. The affected volume reported no available capacity even though the wider storage cluster had ample free space, which made the brokers backing the topic unavailable.

A contributing factor was a startup dependency in the event-processing service: the event lifecycle publishing path is an auxiliary path, but a failure to connect to it prevented the service from starting at all. This turned the loss of a single non-critical topic into a full event-processing outage.

> We had also not yet completed monitoring and alerting for capacity and health of this new storage backend, which meant the failure surfaced through its downstream impact rather than ahead of it.

Reviewer: Rephrase to less "we" and more facts:

```suggestion
Full monitoring and alerting for capacity and health of this new Ceph based storage backend was not completed. This caused the failure to surface via downstream impact rather than at the source.
```
